File: python3/9999.快速排序.py
def quick_sort(array):
    def qsort(nums, start, end):
        if start >= end:
            return nums

        value = nums[start]
        left, right = start, end
        while left < right:
            while left < right and nums[right] <= value:
                right -= 1
            while left < right and nums[left] >= value:
                left += 1
            nums[left], nums[right] = nums[right], nums[left]

        nums[start], nums[left] = nums[left], nums[start]

        qsort(nums, start, left - 1)
        qsort(nums, right + 1, end)

        return nums

    return qsort(array, 0, len(array) - 1)

File: java/9999.快速排序.java
class Solution {
    public static int[] quickSort(int[] array) {
        if (array == null || array.length == 0)
            return array;

        return sort(array, 0, array.length - 1);
    }

    private static int[] sort(int[] array, int start, int end) {
        if (start >= end)
            return array;

        int left = start, right = end;
        int pivot = array[start];

        while (left < right) {
            while (left < right && array[right] >= pivot)
                right--;
            while (left < right && array[left] <= pivot)
                left++;

            swap(array, left, right);
        }

        swap(array, start, left);
        sort(array, start, left - 1);
        sort(array, right + 1, end);

        return array;
    }

    private static void swap(int[] array, int i, int j) {
        int v = array[i];
        array[i] = array[j];
        array[j] = v;
    }
}
